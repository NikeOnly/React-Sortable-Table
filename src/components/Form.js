import React from 'react';
import validation from '../utils/validation';

export default class Main extends React.PureComponent {
    constructor(props) {
        super(props);

        this.state = this.getInitState()
    }

    getInitState() {
        return {
            firstName: '',
            firstNameError: false,
            lastName: '',
            lastNameError: false,
            phone: '',
            phoneError: false,
            gender: 'Male',
            age: '',
            ageError: false,

            isValidationDone: ''
        }
    }

    handleChange(e, field) {
        const errorMessage = validation(e.target.value, field)
        const fieldError = field + 'Error';

        this.setState({ [field]: e.target.value, [fieldError]: errorMessage })
    }

    handleBlur(e, field) {
        const errorMessage = validation(e.target.value, field)
        const fieldError = field + 'Error';

        this.setState({ [fieldError]: errorMessage })
    }

    onAddField(e) {
        e.preventDefault();

        if(this.state.firstNameError === '' && this.state.lastNameError === '' && this.state.phoneError === '' && this.state.ageError === '') {

            this.props.addField(
                this.state.firstName,
                this.state.lastName,
                this.state.phone,
                this.state.gender,
                this.state.age
            );

            this.setState({ ...this.getInitState() })
        } else {
            this.setState({ isValidationDone: 'Please, fill all fields correctly!' })
            setTimeout(() => {
                this.setState({ isValidationDone: '' })
            }, 2500)
        }
    }

    render() {
        const {firstName, firstNameError, lastName, lastNameError, phone, phoneError, gender, age, ageError,
            isDisabled} = this.state;

        return (
            <form onSubmit={this.onAddField.bind(this)}>
                <div className={`form-group ${firstNameError ? 'has-error' : '' }`}>
                    <label htmlFor="firstName">First Name</label>
                    <input type="text"
                        className="form-control"
                        id="firstName"
                        placeholder="First Name"
                        value={firstName}
                        onChange={e => this.handleChange(e, 'firstName')}
                        onBlur={e => this.handleBlur(e, 'firstName')}
                    />
                    <span className="help-block">{firstNameError}</span>
                </div>
                <div className={`form-group ${lastNameError ? 'has-error' : '' }`}>
                    <label htmlFor="lastName">Last Name</label>
                    <input type="text"
                        className="form-control"
                        id="lastName"
                        placeholder="Last Name"
                        value={lastName}
                        onChange={e => this.handleChange(e, 'lastName')}
                        onBlur={e => this.handleBlur(e, 'lastName')}
                    />
                    <span className="help-block">{lastNameError}</span>
                </div>
                <div className={`form-group ${phoneError ? 'has-error' : '' }`}>
                    <label htmlFor="phone">Phone</label>
                    <input type="tel"
                        className="form-control"
                        id="phone"
                        placeholder="Phone"
                        value={phone}
                        onChange={e => this.handleChange(e, 'phone')}
                        onBlur={e => this.handleBlur(e, 'phone')}
                    />
                    <span className="help-block">{phoneError}</span>
                </div>
                <div className="form-group">
                    <label className="radio-inline">
                      <input type="radio"
                          value="Male"
                          onChange={e => this.handleChange(e, 'gender')}
                          checked={gender === 'Male'}
                      />
                      Male
                    </label>
                    <label className="radio-inline">
                      <input type="radio"
                          value="Female"
                          onChange={e => this.handleChange(e, 'gender')}
                          checked={gender === 'Female'}
                      />
                      Female
                    </label>
                </div>
                <div className={`form-group ${ageError ? 'has-error' : '' }`}>
                    <label htmlFor="age">Age</label>
                    <input type="number"
                        className="form-control"
                        id="age"
                        placeholder="Age"
                        value={age}
                        onChange={e => this.handleChange(e, 'age')}
                        onBlur={e => this.handleBlur(e, 'age')}
                    />
                    <span className="help-block">{ageError}</span>
                </div>
                <button type="submit" className="btn btn-default">Submit</button>
                <div className="has-error">
                    <span className="help-block">{this.state.isValidationDone}</span>
                </div>
            </form>
        );
    }
}
