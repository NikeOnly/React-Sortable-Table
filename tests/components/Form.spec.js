import React from 'react'
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme'

import Form from '../../src/components/Form';

describe('Form', () => {
    test('match snapshot', () => {
        const tree = renderer.create(<Form />).toJSON()
        expect(tree).toMatchSnapshot();
    })

    describe('input fields and fieldErrors get correct values', () => {
        const wrapper = mount(<Form />);
        let inputs = wrapper.find('.form-control');
        let radios = wrapper.find({type: 'radio'});

        test('inputs handleChange, onBlur & validation functions', () => {
            inputs.forEach(input => {
                const id = input.prop('id');
                const error = id + 'Error';

                if (id === 'firstName' || id === 'lastName') {
                    input.props().onChange({target: {value: '1'}}, id)
                    input.props().onBlur({target: {value: '1'}}, id)
                    expect(wrapper.state(id)).toBe('1')
                    expect(typeof( wrapper.state(error) )).toBe('string')

                    input.props().onChange({target: {value: 'test'}}, id)
                    input.props().onBlur({target: {value: 'test'}}, id)
                    expect(wrapper.state(id)).toBe('test')
                    expect(wrapper.state(error)).toBeFalsy()
                }

                if (id === 'phone' || id === 'age') {
                    input.props().onChange({target: {value: '1'}}, id)
                    input.props().onBlur({target: {value: '1'}}, id)
                    expect(wrapper.state(id)).toBe('1')
                    expect(wrapper.state(error)).toBeFalsy()

                    input.props().onChange({target: {value: 'test'}}, id)
                    input.props().onBlur({target: {value: 'test'}}, id)
                    expect(wrapper.state(id)).toBe('test')
                    expect(typeof( wrapper.state(error) )).toBe('string')
                }

                if (id === 'age') {
                    input.props().onChange({target: {value: ''}}, id)
                    expect(typeof( wrapper.state(error) )).toBe('string')

                    input.props().onChange({target: {value: '300'}}, id)
                    expect(typeof( wrapper.state(error) )).toBe('string')
                }
            })
        })

        test('radio handleChange function', () => {
            radios.forEach(radio => {
                const value = radio.prop('value');

                radio.props().onChange({target: {value}}, 'gender')

                expect(radio.prop('checked')).toBeTruthy()
                expect(wrapper.state('gender')).toEqual(value)
            })
        })
    })

    test('addField', (done) => {
        const addField = jest.fn();
        const wrapper = mount(<Form addField={addField} />);
        const form = wrapper.find('form');

        form.simulate('submit', { preventDefault: () => {} })

        expect(typeof( wrapper.state('isValidationDone') )).toBe('string')

        setTimeout(() => {
            expect(wrapper.state('isValidationDone')).toEqual('')
            done();
        }, 2500);        

        wrapper.setState({firstNameError: '', lastNameError: '', phoneError: '', ageError: ''})

        form.simulate('submit', { preventDefault: () => {} })

        expect(wrapper.state('isValidationDone')).toEqual('')
        expect(wrapper.prop('addField')).toHaveBeenCalled()
    })
})
