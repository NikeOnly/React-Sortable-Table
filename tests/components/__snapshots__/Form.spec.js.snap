// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form match snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group "
  >
    <label
      htmlFor="firstName"
    >
      First Name
    </label>
    <input
      className="form-control"
      id="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="First Name"
      type="text"
      value=""
    />
    <span
      className="help-block"
    />
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="lastName"
    >
      Last Name
    </label>
    <input
      className="form-control"
      id="lastName"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Last Name"
      type="text"
      value=""
    />
    <span
      className="help-block"
    />
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="phone"
    >
      Phone
    </label>
    <input
      className="form-control"
      id="phone"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Phone"
      type="tel"
      value=""
    />
    <span
      className="help-block"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="radio-inline"
    >
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value="Male"
      />
      Male
    </label>
    <label
      className="radio-inline"
    >
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value="Female"
      />
      Female
    </label>
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="age"
    >
      Age
    </label>
    <input
      className="form-control"
      id="age"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Age"
      type="number"
      value=""
    />
    <span
      className="help-block"
    />
  </div>
  <button
    className="btn btn-default"
    type="submit"
  >
    Submit
  </button>
  <div
    className="has-error"
  >
    <span
      className="help-block"
    >
      
    </span>
  </div>
</form>
`;
